Core/src/org/sleuthkit/autopsy/casemodule/CollaborationMonitor.java

```diff
@@ -0,0 +1,700 @@
+                    progress.finish();
+                });
+                taskIdsToProgressBars.clear();
+            }
+
+            /**
+             * Determines whether or not the time since the last update of this
+             * remote tasksForHost collection is greater than the maximum
+             * acceptable interval between updates.
+             *
+             * @return True or false.
+             */
+            boolean isStale() {
+                return Duration.between(lastUpdateTime, Instant.now()).toMinutes() >= MAX_MINUTES_WITHOUT_UPDATE;
+            }
+        }
+
+    }
+
+    /**
+     * A Runnable task that periodically publishes the local tasksForHost in
+     * progress on this node, providing a heartbeat message for collaboration
+     * monitors on other nodes. The current local tasksForHost are included in
+     * the heartbeat message so that nodes that have just joined the event
+     * channel know what this node is doing, even if they join after the current
+     * tasksForHost are begun.
+     */
+    private final class HeartbeatTask implements Runnable {
+
+        /**
+         * Publish a heartbeat message.
+         */
+        @Override
+        public void run() {
+            eventPublisher.publishRemotely(new CollaborationEvent(hostName, localTasksManager.getCurrentTasks()));
+        }
+    }
+
+    /**
+     * A Runnable task that periodically deals with any remote tasksForHost that
+     * have gone stale, i.e., tasksForHost for which updates have ceased,
+     * presumably because the collaborating node has gone down or there is a
+     * network issue.
+     */
+    private final class StaleTaskDetectionTask implements Runnable {
+
+        /**
+         * Check for stale remote tasksForHost and clean them up, if found.
+         */
+        @Override
+        public void run() {
+            remoteTasksManager.finishStaleTasks();
+        }
+    }
+
+    /**
+     * A Runnable task that periodically checks the availability of
+     * collaboration resources (PostgreSQL server, Solr server, Active MQ
+     * message broker) and reports status to the user in case of a gap in
+     * service.
+     */
+    private final static class CrashDetectionTask implements Runnable {
+
+        private static boolean dbServerIsRunning = true;
+        private static boolean solrServerIsRunning = true;
+        private static boolean messageServerIsRunning = true;
+        private static final Object lock = new Object();
+
+        /**
+         * Monitor the availability of collaboration resources
+         */
+        @Override
+        public void run() {
+            synchronized (lock) {
+                CaseDbConnectionInfo dbInfo = UserPreferences.getDatabaseConnectionInfo();
+                try {
+                    DriverManager.getConnection("jdbc:postgresql://" + dbInfo.getHost() + ":" + dbInfo.getPort() + "/" + "postgres", dbInfo.getUserName(), dbInfo.getUserName()); // NON-NLS
```

Comment:
Yeah. We probably don't want to expose the connection pool like that. I was
trying to indicate settingsValid(). A different name may help.

In the pull request that is sitting out there for connection pools, it does
close the connection. I added the close() methods, I think everywhere
(ignoring SQLite in imagegallery and timeline, of course).

On Tue, Jun 9, 2015 at 11:59 AM, Richard Cordovano <notifications@github.com
> wrote:

> In Core/src/org/sleuthkit/autopsy/casemodule/CollaborationMonitor.java
> <https://github.com/sleuthkit/autopsy/pull/1325#discussion_r32029780>:
>
> > +    private final static class CrashDetectionTask implements Runnable {
> > +
> > +        private static boolean dbServerIsRunning = true;
> > +        private static boolean solrServerIsRunning = true;
> > +        private static boolean messageServerIsRunning = true;
> > +        private static final Object lock = new Object();
> > +
> > +        /**
> > +         * Monitor the availability of collaboration resources
> > +         */
> > +        @Override
> > +        public void run() {
> > +            synchronized (lock) {
> > +                CaseDbConnectionInfo dbInfo = UserPreferences.getDatabaseConnectionInfo();
> > +                try {
> > +                    DriverManager.getConnection("jdbc:postgresql://" + dbInfo.getHost() + ":" + dbInfo.getPort() + "/" + "postgres", dbInfo.getUserName(), dbInfo.getUserName()); // NON-NLS
>
> Yes, it would be nice to use the connection pool for this sort of check.
> But I'm not sure we want to expose connections to the case database in such
> direct fashion, and since this code is connecting to the postgres database,
> connection pooling is probably not needed. Perhaps there could be a test
> method on SleuthkitCase, and we could change the test in the collaboration
> monitor.
>
> Setting that idea aside for now, I'm not sure how line 524 and 526 are
> doing almost the same thing. UserPreferences just constructs a
> CaseDbConnectinoInfo. However, I do see that CaseDbConnectinoInfo has a
> settingsValid() method that does the same sort of thing, so calling that
> would be better than duplicating code, is that what you are suggesting? If
> so, sounds good to me, but the method name settingsValid() and the docs
> don't really make it clear that the validation mechanism is making a
> connection - it could just be checking the form of the settings, for
> example. We should probably change the name of settingsValid() to something
> like canConnect(). And CaseDbConnectinoInfo.settingsValid() is also not
> closing the connection, so we will need to fix that.
>
> As for the copy-paste error, thanks, that slipped by my testing because I
> was using the default user name and password, which are the same. Good
> catch!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sleuthkit/autopsy/pull/1325/files#r32029780>.
>